### ISSUE TITLE
Eslint 9

### ISSUE THREAD

**Issue by mrgrain** [projen/projen#3240]
Eslint 9 release is looming and it comes with a whole bunch of breaking changes: https://eslint.org/docs/next/use/migrate-to-9.0.0

The biggest ones from projen's perspective are:
- [New default config format](https://eslint.org/docs/next/use/migrate-to-9.0.0#-new-default-config-format-eslintconfigjs)
- [Deprecation of formatting rules](https://eslint.org/blog/2023/10/deprecating-formatting-rules/)

I think we will need a
- A new component for basic JS config files (see https://github.com/projen/projen/pull/2414 and https://github.com/projen/projen/issues/2405)
- A new component `Eslint9` or `EslintModern`or whatever
  - Should use the new js config file component
  - Require a correct version of eslint
  - Make this component the default unless a `useLegacy` option is set
  - Enable prettier by default for formatting. 
- Deprecate and warn in the old component

**Comment by rjmackay** [comment]
Eslint 8 is now EOL. Any updates / ETA on updating?

**Comment by mrgrain** [comment]
> Eslint 8 is now EOL. Any updates / ETA on updating?

Not really. Looks like Eslint 9 still supports the old config format though. So that might be a reasonable stop-gap.

**Comment by mrgrain** [comment]
Basic Eslint 9 support landed via #3935 however this doesn't cover the new config style yet.

**Comment by mrgrain** [comment]
Some thoughts based on previous attempts:
 
- The size of the added surface to the public API should be kept minmal. I understand the urge to provide something foundational and flexible. Instead I suggest we focus on a very narrow public API for a JavaScript data/config file.
- What are the must have requirements for that? Surely we won't need the whole JS feature set to write an eslint config. We can start by converting the existing default configs and see what features we need.
- Also focus on the DX. I feel it should be closed to how other object files are defined. They have a really straight forward API and still support lazy-resolution via arrow functions. 
- Ensure we include a new documentation page covering  JS data/config files. Wwe don't have a section for this yet, there probably should be one covering all sorts of files.
- The current proposals seems to be primarily adding two features:  a/ strings with tokens, and b/ conversion from JS literals to valid JavaScript syntax as string. If my understanding is correct here, both might be needed but can and should be implemented independently of each other. E.g. tokenized strings are something that should exist in its own without an inherent connection to (JS) code. We might not really need tokenized strings if get the API right though.

**Comment by dsdavis4** [comment]
I am noticing a lot of activity across multiple issues/prs regarding eslint 9 updates. Specifically a full migration to eslint 9 support with the new config. A few links listed below. Is there a centralized place where discussions on these efforts are taking place? I am wondering if this is on the roadmap and a potential timeline. Additionally, are there efforts I could do to help move it along? 

https://github.com/projen/projen/pull/3454
https://github.com/projen/projen/issues/3950
https://github.com/projen/projen/pull/3711

**Comment by mrgrain** [comment]
I reckon this issue should be the central place. Thanks for linking the other discussions.
